chain/chain/src/chain.rs

```diff
@@ -2407,6 +2414,84 @@ impl Chain {
         )
     }
 
+    fn get_epoch_start_sync_hash_impl(
+        &self,
+        sync_header: &BlockHeader,
+    ) -> Result<CryptoHash, Error> {
+        let mut epoch_id = *sync_header.epoch_id();
+        let mut hash = *sync_header.hash();
+        let mut prev_hash = *sync_header.prev_hash();
+        loop {
+            if prev_hash == CryptoHash::default() {
+                return Ok(hash);
+            }
+            let header = self.get_block_header(&prev_hash)?;
+            if &epoch_id != header.epoch_id() {
+                return Ok(hash);
+            }
+            epoch_id = *header.epoch_id();
+            hash = *header.hash();
+            prev_hash = *header.prev_hash();
+        }
+    }
+
+    /// Find the hash of the first block on the same epoch (and chain) of block with hash `sync_hash`.
+    pub fn get_epoch_start_sync_hash(&self, sync_hash: &CryptoHash) -> Result<CryptoHash, Error> {
+        let header = self.get_block_header(sync_hash)?;
+        self.get_epoch_start_sync_hash_impl(&header)
+    }
+
+    // Returns the first block for which at least two new chunks have been produced for every shard in the epoch
+    pub fn get_current_epoch_sync_hash(
+        &self,
+        sync_header: &BlockHeader,
```

Comment:
This part is because actually if you just pass a sync_hash to it, it will fail in get_block_header() because this function is called in start_process_block_impl() on a block header that has not yet been saved. So if you call get_epoch_start_sync_hash  on this block, it will fail. But we do actually have the block header from the caller, so we can just pass it as an argument even though it’s not saved to the DB yet

Reply:
Ah I think it's just the naming that confused me, can you clarify what is the sync header and / or rename it appropriately?